components/style/properties/properties.mako.rs

```diff
@@ -1003,44 +1064,78 @@ impl UnparsedValue {
+                None => unreachable!("Not from shorthand?"),
+            };
+            let is_valid = self.substitute_variables_and_parse(
+                custom_properties,
+                quirks_mode,
+                |context, input| match shorthand_id {
+                    // No need to parse the 'all' shorthand as anything other than
+                    // a CSS-wide keyword, after variable substitution.
+                    ShorthandId::All => {
+                        Err(input.new_custom_error(StyleParseErrorKind::UnspecifiedError))
                     }
                     % for shorthand in data.shorthands_except_all():
-                        Some(ShorthandId::${shorthand.camel_case}) => {
-                            shorthands::${shorthand.ident}::parse_value(&context, input)
-                            .map(|longhands| {
-                                match longhand_id {
-                                    % for property in shorthand.sub_properties:
-                                        LonghandId::${property.camel_case} => {
-                                            PropertyDeclaration::${property.camel_case}(
-                                                longhands.${property.ident}
-                                            )
-                                        }
-                                    % endfor
-                                    _ => unreachable!()
-                                }
-                            })
-                        }
+                    ShorthandId::${shorthand.camel_case} => {
+                        cache.unparsed = Some(self);
+                        cache.decls.clear();
+                        shorthands::${shorthand.ident}::
+                            parse_into(&mut cache.decls, &context, input)
+                    }
                     % endfor
                 }
-            })
-            .ok()
-        })
-        .unwrap_or_else(|| {
-            // Invalid at computed-value time.
-            let keyword = if longhand_id.inherited() {
-                CSSWideKeyword::Inherit
-            } else {
-                CSSWideKeyword::Initial
-            };
-            PropertyDeclaration::CSSWideKeyword(longhand_id, keyword)
-        })
+            ).is_some();
+            cache.is_valid = is_valid;
+        }
+        if !cache.is_valid {
+            return None;
+        }
+        let mut idx = None;
+        // In cascading, we iterate the declarations in reverse order,
+        // while subproperties of shorthand are added to declaration
+        // blocks in normal order. We use reverse order here as well,
+        // so that in common cases, each item would be visited only
+        // once. Note that not all callsites of the function does the
+        // same thing as cascading. We may want to change them if their
+        // performance matters.
+        for (i, decl) in cache.decls.declarations.iter().enumerate().rev() {
```

Comment:
nit: this is just `position()`.

Reply:
Oh, yes. `rposition()` to be exact, though :)